File: 程序员面试金典/4.8首个共同祖先/lowestCommonAncestor.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        p_path = []
        self.find_path(root, p, p_path)
        p_path.reverse()
        q_path = []
        self.find_path(root, q, q_path)
        q_path.reverse()
        i = 0
        while i < len(p_path) and i < len(q_path):
            if p_path[i] is not q_path[i]:
                break
            i += 1
        return p_path[i - 1]
    
    def find_path(self, root, node, path):
        if root is None:
            return False
        elif root is node or self.find_path(root.left, node, path) or self.find_path(root.right, node, path):
            path.append(root)
            return True
        else:
            return False

File: 程序员面试金典/4.8/lowestCommonAncestor.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

// 1253 ms
class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if (root == null || root == p || root == q)
            return root;
        boolean pIsOnLeft = covers(root.left, p);
        boolean qIsOnLeft = covers(root.left, q);
        if (pIsOnLeft != qIsOnLeft)
            return root;
        return lowestCommonAncestor(pIsOnLeft ? root.left : root.right, p, q);
    }

    private boolean covers(TreeNode root, TreeNode p) {
        if (root == null)
            return false;
        if (root == p)
            return true;
        return covers(root.left, p) || covers(root.right, p);
    }

}
